google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineDeployAction.java

```diff
@@ -0,0 +1,248 @@
+      stagingDirectory = FileUtil.createTempDirectory(
+          "gae-mvm" /* prefix */,
+          null /* suffix */,
+          true  /* deleteOnExit */);
+      consoleLogLn(
+          "Created temporary staging directory: " + stagingDirectory.getAbsolutePath());
+    } catch (IOException e) {
+      logger.error(e);
+      callback.errorOccurred(
+          GctBundle.message("appengine.deployment.error.creating.staging.directory"));
+      return;
+    }
+
+    try {
+      File stagedArtifactPath =
+          copyFile(stagingDirectory, "target" + artifactType, deploymentArtifactPath);
+      stagedArtifactPath.setReadable(true /* readable */, false /* ownerOnly */);
+
+      if (this.appYamlPath != null) {
+        copyFile(stagingDirectory, "app.yaml", this.appYamlPath);
+      }
+      if (this.dockerFilePath != null) {
+        copyFile(stagingDirectory, "Dockerfile", this.dockerFilePath);
+      }
+    } catch (IOException e) {
+      logger.error(e);
+      callback.errorOccurred(GctBundle.message("appengine.deployment.error.during.staging"));
+      return;
+    }
+
+    String version = AppEngineUtil.generateVersion();
+
+    GeneralCommandLine commandLine = new GeneralCommandLine(
+        appEngineHelper.getGcloudCommandPath().getAbsolutePath());
+    commandLine.addParameters("preview", "app", "deploy", "app.yaml", "--promote", "--quiet");
+    commandLine.addParameter("--version=" + version);
+    commandLine.addParameter("--format=json");
+
+    commandLine.withWorkDirectory(stagingDirectory);
+    consoleLogLn("Working directory set to: " + stagingDirectory.getAbsolutePath());
+
+    try {
+      executeProcess(
+          commandLine,
+          new DeployToAppEngineProcessListener(version));
+    } catch (ExecutionException e) {
+      logger.error(e);
+      callback.errorOccurred(GctBundle.message("appengine.deployment.error.during.execution"));
+    }
+  }
+
+  private File copyFile(File stagingDirectory, String targetFileName, File sourceFilePath)
+      throws IOException {
+    File destinationFilePath = new File(stagingDirectory, targetFileName);
+    FileUtil.copy(sourceFilePath, destinationFilePath);
+    consoleLogLn("Copied %s %s to %s", targetFileName,
+        sourceFilePath.getAbsolutePath(), destinationFilePath.getAbsolutePath());
+    return destinationFilePath;
+  }
+
+  private class DeployToAppEngineProcessListener extends ProcessAdapter {
+    private StringBuilder deployOutput = new StringBuilder();
+    private String version;
+
+    public DeployToAppEngineProcessListener(String version) {
+      this.version = version;
+    }
+
+    @Override
+    public void onTextAvailable(ProcessEvent event, Key outputType) {
+      if(outputType.equals(ProcessOutputTypes.STDOUT)) {
+        deployOutput.append(event.getText());
+      }
+    }
+
+    @Override
+    public void processTerminated(final ProcessEvent event) {
+      if (event.getExitCode() == 0) {
+        callback.succeeded(new DeploymentRuntime() {
```

Comment:
can we define this DeploymentRuntime somewhere else?  It distracts from the flow.